wordpress-mobile/WordPress-Android
# My Site Dashboard: Page Behaviours - Error States - Part 1

Parent #15625

##### PS: PR size increase is due to a lot of boilerplate code that is added to introduce a new adapter + view holder.

Page Behavior (pbArwn-3ny-p2)

This PR
- Adds a dedicated dashboard cards builder to allow adding a new dashboard error card and grouping all existing and any future dashboard cards together. 
- Converts post card to a dashboard card.
- Adds a dashboard error card (yet to be displayed).

To test:

- As the newly added dashboard error card is not yet displayed, it'll be sufficient to check that the newly added dashboard cards builder tests include the needed tests.

- Also, it'll be useful if you can take a look at the updated data structure and building logic for the dashboard cards. 

Notes:
1. To prematurely see the dashboard error card, 
    - Add below line at [`line 263`](https://github.com/wordpress-mobile/WordPress-Android/compare/issue/15625-page-behaviours-error-state-part-i?expand=1#diff-74b60070be4a7d0e6d9939f75161ca97a71b1c072489e04c4c3a790699cee910R263) inside `MySiteViewModel.DashboardCardsBuilderParams`:
    ` showErrorCard = true,`
   - Then, install the app and 
      - Go to `My Site` -> `Me` -> `App Settings` -> `Test Feature Configuration`.
      - Turn the `MySiteDashboardPhase2FeatureConfig` feature flag ON.
      - Relaunch the app.
    - Then, notice that dashboard error card is displayed above the site items on `My Site` tab.
       
      ![Dashboard_Card](https://user-images.githubusercontent.com/1405144/146887732-f30ffeeb-3251-44b6-bbf2-941b50338625.png)

2. Ignore styling for the dashboard error card. Note that it doesn't include the retry button yet.

## Regression Notes
1. Potential unintended areas of impact
- Existing cards should build as before.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Updated existing unit tests.

3. What automated tests I added (or what prevented me from doing so)
- Added new unit tests for the newly added dashboard cards builder.

PR submission checklist:
